src/gradio_ui.py

```diff
@@ -100,6 +100,46 @@ def _regenerate(
         def _undo_last_exchange(model_history, gradio_history):
             '''Undoes the last exchange (message pair).'''
             return model_history[:-2], gradio_history[:-1], gradio_history[:-1]
+            
+        def _save_chat_history(gradio_history, *char_setting_states):
+            '''Saves the current chat history to a .json file.'''
+            char_name = char_setting_states[0]
+            with open(f"{char_name}_conversation.json", "w") as f:
+                f.write(json.dumps({"chat": gradio_history}))
+            return f"{char_name}_conversation.json"
+            
+                
+        def _load_chat_history(file_obj, *char_setting_states):
+            '''Loads up a chat history from a .json file.'''
+            file_data = json.loads(file_obj.decode('utf-8'))
+            gradio_history = file_data["chat"]
+            char_name = char_setting_states[0]
+            
+            # If the chat is more than 4 turns we grab only the last 4 for the model history
+            model_history = gradio_history[-4:] if len(gradio_history) >= 4 else gradio_history
+            # Format the gradio history into a new model history
+            new_model_history = []
+            for line in itertools.chain.from_iterable(model_history):
+                # Chop off "\n" from the end of the line
+                formatted_line = line[:-1]
+                # Replace <em> with asterisks
+                formatted_line = formatted_line.replace("<em>", "*").replace("</em>", "*")
+                # Remove <p> and <strong>
+                formatted_line = formatted_line.replace("<p>", "").replace("</p>", "")
+                formatted_line = formatted_line.replace("<strong>", "").replace("</strong>", "")
+                
+                # If there's no indicator of the speaker, it's (You) who's speaking
+                speaker_prefix = formatted_line.split(":")[0]
+                if speaker_prefix != char_name:
+                    formatted_line = f"You: {formatted_line}"
+                # Account for the scenario where the user loads in the conversation first
+                # before loading in the character data
+                elif char_name == "" and speaker_prefix != "":
+                    char_name = speaker_prefix
```

Comment:
Is there any reason we're saving the Gradio history instead of the model history? If we handled the model history instead, we could skip most of this since Gradio takes care of it automatically.

Reply:
From what I understand, the full Gradio history contains the entire chat that was sent in the chatbox. Unless I understood it wrong, model_history is only the memory of the model. Unless I misunderstood what model_history was.

Reply:
`model_history` is the entire chat, as the model should see it (asterisks for italics instead of HTML tags and so on). It's equivalent to the contents of `gradio_history`, just without all the stuff Gradio adds automatically - which is what you're having to strip out here.

Reply:
Ah. I see. Fixed code for it